### PR TITLE
AP_Periph: added serial options support

### PR DESCRIPTION
**Pull request by tridge**
This makes it possible to have a DroneCAN serial port with options such as swap TX/RX, inversion, half-duplex etc
It also allows for flow control settings

### CODE REVIEW

File: Tools/AP_Periph/serial_options_dev.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    // @Param: OPTIONS
+    // @DisplayName: Serial options
+    // @Description: Control over UART options. The InvertRX option controls invert of the receive pin. The InvertTX option controls invert of the transmit pin. The HalfDuplex option controls half-duplex (onewire) mode, where both transmit and receive is done on the transmit wire. The Swap option allows the RX and TX pins to be swapped on STM32F7 based boards.
+    // @Bitmask: 0:InvertRX, 1:InvertTX, 2:HalfDuplex, 3:SwapTXRX, 4: RX_PullDown, 5: RX_PullUp, 6: TX_PullDown, 7: TX_PullUp, 8: RX_NoDMA, 9: TX_NoDMA, 10: Don't forward mavlink to/from, 11: DisableFIFO, 12: Ignore Streamrate
+    AP_GROUPINFO("OPTIONS", 1, SerialOptionsDev, options, 0),
+
+    // @Param: RTSCTS
+    // @DisplayName: Serial1 flow control
+    // @Description: Enable flow control on serial 2 (telemetry 2). You must have the RTS and CTS pins available on the port.If this is set to 2 then flow control will be auto-detected by checking for the output buffer filling on startup.
```
Comment (magicrub) [inline Tools/AP_Periph/serial_options_dev.cpp:40]: Remove serial-2 specific wording

File: Tools/AP_Periph/serial_options_dev.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+const AP_Param::GroupInfo SerialOptionsDev::var_info[] {
+
+    // @Param: OPTIONS
+    // @DisplayName: Serial options
+    // @Description: Control over UART options. The InvertRX option controls invert of the receive pin. The InvertTX option controls invert of the transmit pin. The HalfDuplex option controls half-duplex (onewire) mode, where both transmit and receive is done on the transmit wire. The Swap option allows the RX and TX pins to be swapped on STM32F7 based boards.
+    // @Bitmask: 0:InvertRX, 1:InvertTX, 2:HalfDuplex, 3:SwapTXRX, 4: RX_PullDown, 5: RX_PullUp, 6: TX_PullDown, 7: TX_PullUp, 8: RX_NoDMA, 9: TX_NoDMA, 10: Don't forward mavlink to/from, 11: DisableFIFO, 12: Ignore Streamrate
+    AP_GROUPINFO("OPTIONS", 1, SerialOptionsDev, options, 0),
+
+    // @Param: RTSCTS
```
Comment (magicrub) [inline Tools/AP_Periph/serial_options_dev.cpp:38]: Should this be wrapped in HAL_HAVE_RTSCTS_SERIALn somehow? We can't distinguish serial port numbers here but this assumes all ports have the option which is adding param spam

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py [hunk 3]
```diff
@@ -2430,8 +2430,10 @@ def embed_bootloader(self, f):
 
         bp = self.bootloader_path()
         if not os.path.exists(bp):
-            self.error("Bootloader (%s) does not exist and AP_BOOTLOADER_FLASHING_ENABLED" %
-                       (bp,))
+            self.error('''Bootloader (%s) does not exist and AP_BOOTLOADER_FLASHING_ENABLED
+Please run: Tools/scripts/build_bootloaders.py %s
+''' %
+                       (bp,os.path.basename(os.path.dirname(args.hwdef[0]))))
```
Comment (magicrub) [inline libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:2436]: wrong PR?

File: Tools/AP_Periph/serial_options_dev.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    // @DisplayName: Serial1 flow control
+    // @Description: Enable flow control on serial 2 (telemetry 2). You must have the RTS and CTS pins available on the port.If this is set to 2 then flow control will be auto-detected by checking for the output buffer filling on startup.
+    // @Values: 0:Disabled,1:Enabled,2:Auto
+    AP_GROUPINFO("RTSCTS",  2, SerialOptionsDev, rtscts, float(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE)),
```
Comment (magicrub) [inline Tools/AP_Periph/serial_options_dev.cpp:42]: is disable what we want? We use BOARD_SER1_RTSCTS_DEFAULT, which is default 2, for Serial1 and all the other ports have a default of 2.

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py (hunk 3 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py]: i've moved it to a new PR: https://github.com/ArduPilot/ardupilot/pull/25835

File: Tools/AP_Periph/serial_options_dev.cpp (hunk 1 above)
Comment (tridge) [reply Tools/AP_Periph/serial_options_dev.cpp]: fixed

File: Tools/AP_Periph/serial_options_dev.cpp (hunk 2 above)
Comment (tridge) [reply Tools/AP_Periph/serial_options_dev.cpp:38]: we could if we put them all in the top level table, but I think making it easy to expand to new options is worth the exposed parameter

File: Tools/AP_Periph/serial_options_dev.cpp (hunk 4 above)
Comment (tridge) [reply Tools/AP_Periph/serial_options_dev.cpp:42]: atm we don't have any of these nodes with flow control. I think when setting up a node like this you need to think about the flow control (and having the parameters next to each other helps)
I also don't want to be debugging the automatic flow control detection on a periph node

File: Tools/AP_Periph/serial_options.cpp [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+#endif
+
+#if HAL_HAVE_SERIAL1
+    // @Group: 1_
+    // @Path: serial_options_dev.cpp
+    AP_SUBGROUPINFO(devs[1], "1_",  2, SerialOptions, SerialOptionsDev),
+#endif
+
+#if HAL_HAVE_SERIAL2
+    // @Group: SERIAL2_
```
Comment (magicrub) [inline Tools/AP_Periph/serial_options.cpp:43]: Serial2_ should be 2_ shouldn't this be breaking the param checker for name length?

**Review by magicrub — CHANGES_REQUESTED** [review]

File: Tools/AP_Periph/serial_options.cpp (hunk 5 above)
Comment (tridge) [reply Tools/AP_Periph/serial_options.cpp]: thanks! fixed

**Review by magicrub — APPROVED** [review]

**Pull request merged**